qiime/plugin/resource/directory_format.py

```diff
@@ -0,0 +1,177 @@
+        self.pathspec = pathspec
+        self.format = format
+
+    def __get__(self, obj, cls=None):
+        if obj is None:
+            return self
+        return BoundFile(self.name, self.pathspec, self.format, obj)
+
+
+class FileCollection(File):
+    def __init__(self, pathspec, *, format=None):
+        super().__init__(pathspec, format=format)
+        self._path_maker = None
+
+    def set_path_maker(self, function):
+        self._path_maker = function
+        return PathMakerDescriptor(self)
+
+    def __get__(self, obj, cls=None):
+        if obj is None:
+            return self
+
+        if self._path_maker is None:
+            raise NotImplementedError()
+
+        return BoundFileCollection(self.name, self.pathspec, self.format,
+                                   obj, path_maker=self._path_maker)
+
+
+class BoundFile:
+    @property
+    def mode(self):
+        return self._directory_format._mode
+
+    def __init__(self, name, pathspec, format, directory_format):
+        self.name = name
+        self.pathspec = pathspec
+        self.format = format
+        self._directory_format = directory_format
+        self._path_maker = lambda s: pathspec
+
+    def view(self, view_type):
+        from_pattern = transform.ResourcePattern.from_view_type(self.format)
+        to_pattern = transform.ResourcePattern.from_view_type(view_type)
+
+        transformation = from_pattern.make_transformation(to_pattern)
+        return transformation(self.path_maker())
+
+    def set(self, view, view_type, **kwargs):
+        if self.mode != 'w':
+            raise TypeError("Cannot use `set`/`add` when mode=%r" % self.mode)
+        from_pattern = transform.ResourcePattern.from_view_type(view_type)
+        to_pattern = transform.ResourcePattern.from_view_type(self.format)
+
+        transformation = from_pattern.make_transformation(to_pattern)
+        result = transformation(view)
+        result.path._move_or_copy(self.path_maker(**kwargs))
+
+    def _validate_members(self, collected_paths):
+        found_members = False
+        root = pathlib.Path(self._directory_format.path)
+        for path in collected_paths:
+            if re.match(self.pathspec, str(path.relative_to(root))):
+                collected_paths[path] = True
+                found_members = True
+                self.format(path, mode='r').validate()
+        if not found_members:
+            raise ValueError("Missing one or more files for %s: %r"
+                             % (self._directory_format.__class__.__name__,
+                                self.pathspec))
+
+    @property
+    def path_maker(self):
+        def bound_path_maker(**kwargs):
+            # Must wrap in a naive Path, otherwise an OutPath would be summoned
+            # into this world, and would destroy everything in its path.
+            path = pathlib.Path(self._directory_format.path) \
+                / self._path_maker(self._directory_format, **kwargs)
```

Comment:
add comment explaining that `self._directory_format` is bound to `self` making it look like an instance of a different class.